File: python-checks/src/main/java/org/sonar/python/checks/regex/OctalEscapeCheck.java
/*
 * SonarQube Python Plugin
 * Copyright (C) 2011-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.python.checks.regex;

import java.util.Collections;
import org.sonar.check.Rule;
import org.sonar.plugins.python.api.tree.CallExpression;
import org.sonarsource.analyzer.commons.regex.RegexParseResult;
import org.sonarsource.analyzer.commons.regex.ast.CharacterTree;
import org.sonarsource.analyzer.commons.regex.ast.RegexBaseVisitor;

@Rule(key = "S6537")
public class OctalEscapeCheck extends AbstractRegexCheck {

  public static final String MESSAGE = "Consider replacing this octal escape sequence with a Unicode or hexadecimal sequence instead.";

  @Override
  public void checkRegex(RegexParseResult regexParseResult, CallExpression regexFunctionCall) {
    new CharacterFinder().visit(regexParseResult);
  }

  private class CharacterFinder extends RegexBaseVisitor {
    @Override
    public void visitCharacter(CharacterTree tree) {
      if (tree.isEscapeSequence() && tree.getText().matches("\\\\{1,2}\\d+")) {
        addIssue(tree, MESSAGE, null, Collections.emptyList());
      }
      super.visitCharacter(tree);
    }
  }
}

File: python-checks/src/test/resources/checks/regex/octalEscapeCheck.py
import re

def non_compliant():
    re.compile("\\01b") # Noncompliant {{Consider replacing this octal escape sequence with a Unicode or hexadecimal sequence instead.}}
#               ^^^^ 0
    re.compile("\\076b") # Noncompliant {{Consider replacing this octal escape sequence with a Unicode or hexadecimal sequence instead.}}
#               ^^^^^ 0
    re.compile("\\276b") # Noncompliant {{Consider replacing this octal escape sequence with a Unicode or hexadecimal sequence instead.}}
#               ^^^^^ 0
    re.compile("\\2760") # Noncompliant {{Consider replacing this octal escape sequence with a Unicode or hexadecimal sequence instead.}}
#               ^^^^^ 0

def compliant():
    re.compile("\\x01")
    re.compile("(x)\\1")
    re.compile(r"(x)\1")
    re.compile("\\4760") # Back reference
